### PR TITLE
subsys: bluetooth: Refactor callback list in the scan module

### PR DESCRIPTION
**Pull request by ktaborowski**

### CODE REVIEW

**Comment by ktaborowski** [comment]
@grochu , @KAGA164 , please take a look

File: applications/nrf_desktop/src/modules/ble_scan.c [hunk 1]
```diff
@@ -212,6 +212,9 @@ static bool le_param_req(struct bt_conn *conn, struct bt_le_conn_param *param)
 	return true;
 }
 
+BT_SCAN_CB_INIT(scan_cb, scan_filter_match, NULL,
+			scan_connecting_error, scan_connecting);
```
Comment (MarekPieta) [inline applications/nrf_desktop/src/modules/ble_scan.c]: Align (also for other samples)

File: samples/bluetooth/throughput/src/main.c [hunk 2]
```diff
@@ -75,11 +75,8 @@ void scan_connecting_error(struct bt_scan_device_info *device_info)
 	printk("Connecting failed\n");
 }
 
-static struct bt_scan_cb scan_cb = {
-	.filter_match = scan_filter_match,
-	.filter_no_match = scan_filter_no_match,
-	.connecting_error = scan_connecting_error,
-};
+BT_SCAN_CB_INIT(scan_cb, scan_filter_match, NULL,
+			scan_connecting_error, NULL);
```
Comment (MarekPieta) [inline samples/bluetooth/throughput/src/main.c]: Add `scan_filter_no_match `callback too (instead of NULL)

File: include/bluetooth/scan.h [hunk 3]
```diff
@@ -224,6 +224,35 @@ struct bt_scan_device_info {
...
+ *
+ * @param[in] _name Unique name for @ref bt_scan_cb structure.
+ * @param[in] match_fun Scan filter matched function pointer.
+ * @param[in] no_match_fun Scan filter unmatched (the device was not found)
+		 function pointer.
+ * @param[in] error_fun Error when connecting function pointer.
+ * @param[in] connecting_fun Connecting data function pointer.
+ */
+
+#define BT_SCAN_CB_INIT(\
```
Comment (MarekPieta) [inline include/bluetooth/scan.h]: align `\`

**Comment by lemrey** [comment]
Hi @ktaborowski, would it not have been possible to keep the callback pointers in flash just by declaring the `struct bt_scan_cb` as `static const` ?

File: include/bluetooth/scan.h [hunk 4]
```diff
@@ -234,7 +263,19 @@ struct bt_scan_device_info {
  *  an instance, it may be set to NULL and will as a consequence not be
  *  used for that instance.
  */
+
 struct bt_scan_cb {
+	const struct bt_scan_cb_data *cb_addr;
+	sys_snode_t node;
```
Comment (MarekPieta) [inline include/bluetooth/scan.h]: Maybe move `bt_scan_cb `below `bt_scan_cb_data` definition

**Comment by MarekPieta** [comment]
And please fix typos in commit message: `poiter`, `callcack`

**Comment by MarekPieta** [comment]
@lemrey I think we cannot make is constant as long as there is `sys_snode_t ` inside

File: include/bluetooth/scan.h [hunk 5]
```diff
@@ -224,17 +224,41 @@ struct bt_scan_device_info {
...
+ * @param[in] match_fun Scan filter matched function pointer.
+ * @param[in] no_match_fun Scan filter unmatched (the device was not found)
+		 function pointer.
+ * @param[in] error_fun Error when connecting function pointer.
+ * @param[in] connecting_fun Connecting data function pointer.
  */
-struct bt_scan_cb {
+
+#define BT_SCAN_CB_INIT(				\
+		_name,					\
```
Comment (KAGA164) [inline include/bluetooth/scan.h]: Maybe moving _name to previous line and aligment rest of params will be more readable

File: include/bluetooth/scan.h [hunk 6]
```diff
@@ -274,6 +298,21 @@ struct bt_scan_cb {
...
+ *  It is registered with the help of the @ref bt_scan_cb_register() API.
+ *  It's permissible to register multiple instances of this @ref bt_scan_cb
+ *  type, in case different modules of an application are interested in
+ *  tracking the scanning state. If a callback is not of interest for
+ *  an instance, it may be set to NULL and will as a consequence not be
+ *  used for that instance.
+ */
+
+struct bt_scan_cb {
+	const struct bt_scan_cb_data *cb_addr;
```
Comment (joerchan) [inline include/bluetooth/scan.h]: I would suggest renaming to `cb_data` here.

**Comment by thst-nordic** [comment]
There has been changes to the master branch Jenkinsfile.  

in order to get the 'CI/Jenkins/NRF' status you will need to rebase master.

**Comment by ktaborowski** [comment]
All commented changes have been corrected.

File: include/bluetooth/scan.h [hunk 7]
```diff
@@ -292,17 +292,40 @@ struct bt_scan_device_info {
...
+ * @param[in] connecting_fun Connecting data function pointer.
  */
-struct bt_scan_cb {
+
+#define BT_SCAN_CB_INIT(_name,				\
+			match_fun,			\
+			no_match_fun,			\
+			error_fun,			\
+			connecting_fun)			\
+	const struct cb_data _name ## _data = { 	\
```
Comment (pdunaj) [inline include/bluetooth/scan.h]: static

Comment (ktaborowski) [reply include/bluetooth/scan.h]: now this struct is static const

**Comment by lemrey** [comment]
Personally, I don't see much reason to do this. Keeping three pointers in RAM shouldn't be a problem and I liked the non-macro initialization better since it was much simpler, but I won't -1 :)

**Pull request merged**